presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentOptions.java

```diff
@@ -41,8 +42,9 @@
     @Option(name = "--without-presto", title = "without Presto", description = "do not start presto-master")
     public boolean withoutPrestoMaster;
 
+    // to avoid port conflicts, by default don't bind in TestContainers reuse mode
```

Comment:
Multiple envs is very convenient, eg a change which may break many product tests, good to have multiple environments, it's simple to do with small fixes like this, separate networks and separate containers. It's also consistent with what's done in integration tests. Containers are never killed automatically there, you have to kill them manually

Reply:
> Multiple envs is very convenient, 

No doubt about it. However, port binding is also convenient. The best is to have these two items.

Also making environments reusable brings new level of complexity. Adding support for multiple environments makes it even harder. So I would like to limit this.

Reply:
> with testcontainers.reuse.enable=true 

This is a global switch that is required by testcontainers. In my case I have it turned on because I sometimes reuse containers. Most often I don't reuse them. So far it was safe.

> principle of least astonishment

This means to me that product tests behaves in the way I requested them to work. Also one might say that `testcontainers` inside of it is implementation detail.

> Returning to ports, does anyone really need --bind=true? 

Yes, I use them. Either to connect to Presto or to connect 3rd party services like HDFS webui. Also I am not sure if it is not used for debug ports which are also used very frequently.

> which very, very often gives port conflicts.

This is the cost that we accepted. This brings the benefit that I just know that Presto is available at 8080 and I can even hardcode that in some local scripts.

> Under Linux (I think most FOSS developers use Linux),

Notice that macos is also very popular and things there are bit more complicated with docker.

Reply:
For most cases (I tested it, works perfectly) reusing containers is fine. If somebody needs to disable it temporarily, they could use option `--reuse=false` corresponding to `container.withReuse` (by default it would be true)

Similarly, for most cases, random port numbers behave better with reusable containers and everything else on developer machines. If some nasty application needs fixed port number, use `--bind=true`, or even better, something like: `--bind=1234,5678` (list explicitly only those ports)

Local scripts can easily discover port numbers, even if they are written in BASH (`docker inspect -f ...`).

So `--reuse=true` and `--bind=<nothing>` would be defaults (make sense for most people and tests), but they can be changed (`--reuse=false` or `--bind=1234,5678`).